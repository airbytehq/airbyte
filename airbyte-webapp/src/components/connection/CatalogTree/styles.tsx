import styled from "styled-components";

import { Cell } from "components";

export const HeaderCell = styled(Cell)`
  font-size: 10px;
  line-height: 13px;
`;

export const CheckboxCell = styled(HeaderCell)`
  max-width: 55px;
  text-align: center;
`;

export const ArrowCell = styled(HeaderCell)`
  min-width: 40px;
  max-width: 40px;
`;

export const NameContainer = styled.span`
  padding-left: 30px;
`;

export const SyncHeaderContainer = styled.span`
  padding-left: 34px;
  display: inline-block;
  min-width: 62px;
`;

export const SyncCheckboxContainer = styled.div`
  padding-left: 24px;
`;
