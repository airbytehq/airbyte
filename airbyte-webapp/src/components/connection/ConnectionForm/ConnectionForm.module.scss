@use "scss/variables";

.formContainer {
  display: flex;
  flex-direction: column;
  gap: variables.$spacing-md;
}

.readonly {
  pointer-events: none;
}

.leftFieldCol {
  flex: 1;
  max-width: 640px;
  padding-right: variables.$spacing-xl;
}

.rightFieldCol {
  flex: 1;
  max-width: 300px;
}
