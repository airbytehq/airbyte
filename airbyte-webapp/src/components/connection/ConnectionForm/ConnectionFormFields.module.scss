@use "scss/variables";

.formContainer {
  display: flex;
  flex-direction: column;
  gap: variables.$spacing-md;
  position: relative;
}

.readonly {
  pointer-events: none;
}

.leftFieldCol {
  flex: 1;
  max-width: 640px;
  padding-right: 30px;
}

.rightFieldCol {
  flex: 1;
  max-width: 300px;
}

.namespaceFormatLabel {
  flex: 5 0 0;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
}

.tryArrow {
  margin: 0 variables.$spacing-md -1px 0;

  // used to control svg size
  font-size: 14px;
}
