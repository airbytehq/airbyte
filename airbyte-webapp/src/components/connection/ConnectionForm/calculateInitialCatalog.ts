import isEqual from "lodash/isEqual";

import { SyncSchema, SyncSchemaStream } from "core/domain/catalog";
import {
  DestinationSyncMode,
  SyncMode,
  AirbyteStreamConfiguration,
  StreamDescriptor,
  StreamTransform,
} from "core/request/AirbyteClient";

const getDefaultCursorField = (streamNode: SyncSchemaStream): string[] => {
  if (streamNode.stream?.defaultCursorField?.length) {
    return streamNode.stream.defaultCursorField;
  }
  return streamNode.config?.cursorField || [];
};

const clearBreakingFieldChanges = (nodeStream: SyncSchemaStream, breakingChangesByStream: StreamTransform[]) => {
  if (!breakingChangesByStream.length || !nodeStream.config) {
    return nodeStream;
  }

  const { primaryKey, cursorField } = nodeStream.config;

  let clearPrimaryKey = false;
  let clearCursorField = false;

  for (const streamTransformation of breakingChangesByStream) {
    if (!streamTransformation.updateStream || !streamTransformation.updateStream?.length) {
      continue;
    }

    // get all of the removed field paths for this transformation
    const breakingFieldPaths = streamTransformation.updateStream
      .filter(({ breaking }) => breaking)
      .map((update) => update.fieldName);

    // if there is a primary key in the config, and any of its field paths were removed, we'll be clearing it
    if (
      !!primaryKey?.length &&
      primaryKey?.some((primaryKeyPath) => breakingFieldPaths.some((path) => isEqual(primaryKeyPath, path)))
    ) {
      clearPrimaryKey = true;
    }

    // if there is a cursor field, and any of its field path was removed, we'll be clearing it
    if (!!cursorField?.length && breakingFieldPaths.some((path) => isEqual(path, cursorField))) {
      clearCursorField = true;
    }
  }

  if (clearPrimaryKey || clearCursorField) {
    return {
      ...nodeStream,
      config: {
        ...nodeStream.config,
        primaryKey: clearPrimaryKey ? [] : nodeStream.config.primaryKey,
        cursorField: clearCursorField ? [] : nodeStream.config.cursorField,
      },
    };
  }

  return nodeStream;
};

const verifySourceDefinedProperties = (streamNode: SyncSchemaStream, isEditMode: boolean) => {
  if (!streamNode.stream || !streamNode.config || !isEditMode) {
    return streamNode;
  }

  const {
    stream: { sourceDefinedPrimaryKey, sourceDefinedCursor },
  } = streamNode;

  // if there's a source-defined cursor and the mode is correct, set the config to the default
  if (sourceDefinedCursor) {
    streamNode.config.cursorField = streamNode.stream.defaultCursorField;
  }

  // if the primary key doesn't need to be calculated from the source, just return the node
  if (!sourceDefinedPrimaryKey || sourceDefinedPrimaryKey.length === 0) {
    return streamNode;
  }

  // override the primary key with what the source said
  streamNode.config.primaryKey = sourceDefinedPrimaryKey;

  return streamNode;
};

const verifySupportedSyncModes = (streamNode: SyncSchemaStream): SyncSchemaStream => {
  if (!streamNode.stream) {
    return streamNode;
  }
  const {
    stream: { supportedSyncModes },
  } = streamNode;

  if (supportedSyncModes?.length) {
    return streamNode;
  }
  return { ...streamNode, stream: { ...streamNode.stream, supportedSyncModes: [SyncMode.full_refresh] } };
};

const verifyConfigCursorField = (streamNode: SyncSchemaStream): SyncSchemaStream => {
  if (!streamNode.config) {
    return streamNode;
  }
  const { config } = streamNode;

  return {
    ...streamNode,
    config: {
      ...config,
      cursorField: config.cursorField?.length ? config.cursorField : getDefaultCursorField(streamNode),
    },
  };
};

const getOptimalSyncMode = (
  streamNode: SyncSchemaStream,
  supportedDestinationSyncModes: DestinationSyncMode[]
): SyncSchemaStream => {
  const updateStreamConfig = (
    config: Pick<AirbyteStreamConfiguration, "syncMode" | "destinationSyncMode">
  ): SyncSchemaStream => ({
    ...streamNode,
    config: { ...streamNode.config, ...config },
  });
  if (!streamNode.stream?.supportedSyncModes) {
    return streamNode;
  }
  const {
    stream: { supportedSyncModes, sourceDefinedCursor, sourceDefinedPrimaryKey },
  } = streamNode;

  if (
    supportedSyncModes.includes(SyncMode.incremental) &&
    supportedDestinationSyncModes.includes(DestinationSyncMode.append_dedup) &&
    sourceDefinedCursor &&
    sourceDefinedPrimaryKey?.length
  ) {
    return updateStreamConfig({
      syncMode: SyncMode.incremental,
      destinationSyncMode: DestinationSyncMode.append_dedup,
    });
  }

  if (supportedDestinationSyncModes.includes(DestinationSyncMode.overwrite)) {
    return updateStreamConfig({
      syncMode: SyncMode.full_refresh,
      destinationSyncMode: DestinationSyncMode.overwrite,
    });
  }

  if (
    supportedSyncModes.includes(SyncMode.incremental) &&
    supportedDestinationSyncModes.includes(DestinationSyncMode.append)
  ) {
    return updateStreamConfig({
      syncMode: SyncMode.incremental,
      destinationSyncMode: DestinationSyncMode.append,
    });
  }

  if (supportedDestinationSyncModes.includes(DestinationSyncMode.append)) {
    return updateStreamConfig({
      syncMode: SyncMode.full_refresh,
      destinationSyncMode: DestinationSyncMode.append,
    });
  }
  return streamNode;
};

const calculateInitialCatalog = (
  schema: SyncSchema,
  supportedDestinationSyncModes: DestinationSyncMode[],
  streamsWithBreakingFieldChanges?: StreamTransform[],
  isNotCreateMode?: boolean,
  newStreamDescriptors?: StreamDescriptor[]
): SyncSchema => {
  return {
    streams: schema.streams.map<SyncSchemaStream>((apiNode, id) => {
      const nodeWithId: SyncSchemaStream = { ...apiNode, id: id.toString() };
      const nodeStream = verifySourceDefinedProperties(verifySupportedSyncModes(nodeWithId), isNotCreateMode || false);

      // if the stream is new since a refresh, verify cursor and get optimal sync modes
      const isStreamNew = newStreamDescriptors?.some(
        (streamIdFromDiff) =>
          streamIdFromDiff.name === nodeStream.stream?.name &&
          streamIdFromDiff.namespace === nodeStream.stream?.namespace
      );

      // if we're in edit or readonly mode and the stream is not new, check for breaking changes then return
      if (isNotCreateMode && !isStreamNew) {
        // narrow down the breaking field changes from this connection to only those relevant to this stream
        const breakingChangesByStream =
          streamsWithBreakingFieldChanges && streamsWithBreakingFieldChanges.length > 0
            ? streamsWithBreakingFieldChanges.filter(({ streamDescriptor }) => {
                return (
                  streamDescriptor.name === nodeStream.stream?.name &&
                  streamDescriptor.namespace === nodeStream.stream?.namespace
                );
              })
            : [];

        return clearBreakingFieldChanges(nodeStream, breakingChangesByStream);
      }

      return getOptimalSyncMode(verifyConfigCursorField(nodeStream), supportedDestinationSyncModes);
    }),
  };
};

export default calculateInitialCatalog;
