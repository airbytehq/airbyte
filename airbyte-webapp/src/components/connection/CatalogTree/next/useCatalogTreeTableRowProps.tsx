/* eslint-disable css-modules/no-unused-class */
import classNames from "classnames";
import isEqual from "lodash/isEqual";
import { useMemo } from "react";

import { PillButtonVariant } from "components/ui/PillSelect/PillButton";

import { SyncSchemaStream } from "core/domain/catalog";
import { useBulkEditSelect } from "hooks/services/BulkEdit/BulkEditService";
import { useConnectionFormService } from "hooks/services/ConnectionForm/ConnectionFormService";

import styles from "./CatalogTreeTableRow.module.scss";

export type StatusToDisplay = "disabled" | "added" | "removed" | "changed" | "unchanged";

export const useCatalogTreeTableRowProps = (stream: SyncSchemaStream) => {
  const [isSelected] = useBulkEditSelect(stream.id);
  const { initialValues } = useConnectionFormService();

  const isStreamEnabled = stream.config?.selected;

  const statusToDisplay = useMemo<StatusToDisplay>(() => {
    const rowStatusChanged =
      initialValues.syncCatalog.streams.find(
        (item) => item.stream?.name === stream.stream?.name && item.stream?.namespace === stream.stream?.namespace
      )?.config?.selected !== stream.config?.selected;

    const rowChanged = !isEqual(
      initialValues.syncCatalog.streams.find(
        (item) =>
          item.stream &&
          stream.stream &&
          item.stream.name === stream.stream.name &&
          item.stream.namespace === stream.stream.namespace
      )?.config,
      stream.config
    );

    if (!isStreamEnabled && !rowStatusChanged) {
      return "disabled";
    } else if (rowStatusChanged) {
      return isStreamEnabled ? "added" : "removed";
    } else if (rowChanged) {
      return "changed";
    }
    return "unchanged";
  }, [initialValues.syncCatalog.streams, isStreamEnabled, stream.config, stream.stream]);

  const pillButtonVariant = useMemo<PillButtonVariant>(() => {
    if (statusToDisplay === "added" && !isSelected) {
      return "green";
    } else if (statusToDisplay === "removed" && !isSelected) {
      return "red";
    } else if (statusToDisplay === "changed" || isSelected) {
      return "blue";
    }
    return "grey";
  }, [isSelected, statusToDisplay]);

  const streamHeaderContentStyle = classNames(styles.streamHeaderContent, {
    [styles.added]: statusToDisplay === "added" && !isSelected,
    [styles.removed]: statusToDisplay === "removed" && !isSelected,
    [styles.changed]: statusToDisplay === "changed" || isSelected,
    [styles.disabled]: statusToDisplay === "disabled",
  });

  return {
    streamHeaderContentStyle,
    isSelected,
    statusToDisplay,
    pillButtonVariant,
  };
};
