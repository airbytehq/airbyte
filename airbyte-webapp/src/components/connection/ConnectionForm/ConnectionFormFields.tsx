import { faSyncAlt } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import classNames from "classnames";
import { Field, FieldProps } from "formik";
import React from "react";
import { FormattedMessage, useIntl } from "react-intl";
import { useUnmount } from "react-use";

import { ControlLabels } from "components";
import { FormChangeTracker } from "components/common/FormChangeTracker";
import { Button } from "components/ui/Button";
import { FlexContainer } from "components/ui/Flex";
import { Input } from "components/ui/Input";

import { NamespaceDefinitionType } from "core/request/AirbyteClient";
import { useNewTableDesignExperiment } from "hooks/connection/useNewTableDesignExperiment";
import { useConnectionFormService } from "hooks/services/ConnectionForm/ConnectionFormService";
import { FeatureItem, useFeature } from "hooks/services/Feature";
import { useFormChangeTrackerService } from "hooks/services/FormChangeTracker";
import { ValuesProps } from "hooks/services/useConnectionHook";

import styles from "./ConnectionFormFields.module.scss";
import { FormikConnectionFormValues } from "./formConfig";
import { NamespaceDefinitionField } from "./NamespaceDefinitionField";
import { NonBreakingChangesPreferenceField } from "./NonBreakingChangesPreferenceField";
import { useRefreshSourceSchemaWithConfirmationOnDirty } from "./refreshSourceSchemaWithConfirmationOnDirty";
import { ScheduleField } from "./ScheduleField";
import { Section } from "./Section";
import { SyncCatalogField } from "./SyncCatalogField";

interface ConnectionFormFieldsProps {
  values: ValuesProps | FormikConnectionFormValues;
  isSubmitting: boolean;
  dirty: boolean;
}

export const ConnectionFormFields: React.FC<ConnectionFormFieldsProps> = ({ values, isSubmitting, dirty }) => {
  const allowAutoDetectSchema = useFeature(FeatureItem.AllowAutoDetectSchema);

  const { mode, formId } = useConnectionFormService();
  const { formatMessage } = useIntl();
  const { clearFormChange } = useFormChangeTrackerService();

  const readonlyClass = classNames({
    [styles.readonly]: mode === "readonly",
  });

  const refreshSchema = useRefreshSourceSchemaWithConfirmationOnDirty(dirty);

  useUnmount(() => {
    clearFormChange(formId);
  });

  const isNewTableDesignEnabled = useNewTableDesignExperiment();
  const firstSectionTitle = isNewTableDesignEnabled ? undefined : <FormattedMessage id="connection.transfer" />;

  return (
    <>
      {/* FormChangeTracker is here as it has access to everything it needs without being repeated */}
      <FormChangeTracker changed={dirty} formId={formId} />
      <div className={styles.formContainer}>
        <Section title={firstSectionTitle}>
          <ScheduleField />
          {allowAutoDetectSchema && (
            <Field name="nonBreakingChangesPreference" component={NonBreakingChangesPreferenceField} />
          )}
        </Section>
        {!isNewTableDesignEnabled && (
          <Section title={<FormattedMessage id="connection.streams" />}>
            <span className={readonlyClass}>
              <Field name="namespaceDefinition" component={NamespaceDefinitionField} />
            </span>
            {values.namespaceDefinition === NamespaceDefinitionType.customformat && (
              <Field name="namespaceFormat">
                {({ field, meta }: FieldProps<string>) => (
                  <FlexContainer alignItems="center">
                    <div className={styles.leftFieldCol}>
                      <ControlLabels
                        className={styles.namespaceFormatLabel}
                        nextLine
                        error={!!meta.error}
                        label={<FormattedMessage id="connectionForm.namespaceFormat.title" />}
                        infoTooltipContent={<FormattedMessage id="connectionForm.namespaceFormat.subtitle" />}
                      />
                    </div>
                    <div className={classNames(styles.rightFieldCol, readonlyClass)}>
                      <Input
                        {...field}
                        error={!!meta.error}
                        placeholder={formatMessage({
                          id: "connectionForm.namespaceFormat.placeholder",
                        })}
                      />
                    </div>
                  </FlexContainer>
                )}
              </Field>
            )}
            <Field name="prefix">
              {({ field }: FieldProps<string>) => (
                <FlexContainer alignItems="center">
                  <div className={styles.leftFieldCol}>
                    <ControlLabels
                      nextLine
                      optional
                      label={formatMessage({
                        id: "form.prefix",
                      })}
                      infoTooltipContent={formatMessage({
                        id: "form.prefix.message",
                      })}
                    />
                  </div>
                  <div className={styles.rightFieldCol}>
                    <Input
                      {...field}
                      type="text"
                      placeholder={formatMessage({
                        id: `form.prefix.placeholder`,
                      })}
                      data-testid="prefixInput"
                      style={{ pointerEvents: mode === "readonly" ? "none" : "auto" }}
                    />
                  </div>
                </FlexContainer>
              )}
            </Field>
          </Section>
        )}
        <Section flush>
          <Field
            name="syncCatalog.streams"
            component={SyncCatalogField}
            isSubmitting={isSubmitting}
            additionalControl={
              <Button
                onClick={refreshSchema}
                type="button"
                variant="secondary"
                data-testid="refresh-source-schema-btn"
                disabled={isSubmitting}
              >
                <FontAwesomeIcon icon={faSyncAlt} className={styles.tryArrow} />
                <FormattedMessage id="connection.updateSchema" />
              </Button>
            }
          />
        </Section>
      </div>
    </>
  );
};
