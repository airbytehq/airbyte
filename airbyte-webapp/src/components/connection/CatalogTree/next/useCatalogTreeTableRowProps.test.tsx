/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { renderHook } from "@testing-library/react-hooks";
import classNames from "classnames";
import * as formik from "formik";

import { AirbyteStreamAndConfiguration } from "core/request/AirbyteClient";
import * as bulkEditService from "hooks/services/BulkEdit/BulkEditService";
import * as connectionFormService from "hooks/services/ConnectionForm/ConnectionFormService";
import { FormikConnectionFormValues } from "views/Connection/ConnectionForm/formConfig";

// eslint-disable-next-line css-modules/no-unused-class
import styles from "./CatalogTreeTableRow.module.scss";
import { useCatalogTreeTableRowProps } from "./useCatalogTreeTableRowProps";

const mockStream: Partial<AirbyteStreamAndConfiguration> = {
  stream: {
    name: "stream_name",
    namespace: "stream_namespace",
  },
  config: { selected: true, syncMode: "full_refresh", destinationSyncMode: "overwrite" },
};

const mockInitialValues: Partial<FormikConnectionFormValues> = {
  syncCatalog: {
    streams: [
      {
        stream: {
          name: "stream_name",
          namespace: "stream_namespace",
        },
        config: { selected: true, syncMode: "full_refresh", destinationSyncMode: "overwrite" },
      },
    ],
  },
};

describe("<CatalogTreeTableRow />", () => {
  it("should return default styles for a row that starts enabled", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      // eslint-disable-next-line
      return { initialValues: mockInitialValues } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });

    const { result } = renderHook(() => useCatalogTreeTableRowProps(mockStream));

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent));
    expect(result.current.statusIcon).toEqual(null);
    expect(result.current.pillButtonVariant).toEqual("grey");
  });
  it("should return disabled styles for a row that starts disabled", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      return {
        initialValues: {
          syncCatalog: {
            streams: [
              {
                ...mockInitialValues.syncCatalog?.streams[0],
                config: { ...mockInitialValues.syncCatalog?.streams[0].config, selected: false },
              },
            ],
          },
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });
    const { result } = renderHook(() =>
      useCatalogTreeTableRowProps({
        ...mockStream,
        config: { ...mockStream.config!, selected: false },
      })
    );

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.disabled));
    expect(result.current.statusIcon).toEqual(null);
    expect(result.current.pillButtonVariant).toEqual("grey");
  });
  it("should return added styles for a row that is added", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      return {
        initialValues: {
          syncCatalog: {
            streams: [
              {
                ...mockInitialValues.syncCatalog?.streams[0],
                config: { ...mockInitialValues.syncCatalog?.streams[0].config, selected: false },
              },
            ],
          },
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });
    const { result } = renderHook(() =>
      useCatalogTreeTableRowProps({
        ...mockStream,
        config: { ...mockStream.config!, selected: true }, // selected true
      })
    );

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.added));
    // we care that iconName="plus" and className="icon plus" but it is tricky to test a react node returned from a hook like this
    expect(result.current.statusIcon).toMatchSnapshot();
    expect(result.current.pillButtonVariant).toEqual("green");
  });
  it("should return removed styles for a row that is removed", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      return {
        initialValues: { ...mockInitialValues },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });
    const { result } = renderHook(() =>
      useCatalogTreeTableRowProps({
        ...mockStream,
        config: { ...mockStream.config!, selected: false }, // selected false
      })
    );

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.removed));

    // we care that iconName="minus" and className="icon minus" but it is tricky to test a react node returned from a hook like this
    expect(result.current.statusIcon).toMatchSnapshot();
    expect(result.current.pillButtonVariant).toEqual("red");
  });
  it("should return updated styles for a row that is updated", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      return {
        initialValues: { ...mockInitialValues },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });
    const { result } = renderHook(() =>
      useCatalogTreeTableRowProps({
        ...mockStream,
        config: { ...mockStream.config!, syncMode: "incremental", destinationSyncMode: "append_dedup" }, // new sync mode and destination sync mode
      })
    );

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.changed));
    // we care that this is our custom ModificationIcon component with modificationColor as its color
    expect(result.current.statusIcon).toMatchSnapshot();
    expect(result.current.pillButtonVariant).toEqual("blue");
  });
  it("should return added styles for a row that is both added and updated", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      return {
        initialValues: {
          syncCatalog: {
            streams: [
              {
                ...mockInitialValues.syncCatalog?.streams[0],
                config: { ...mockInitialValues.syncCatalog?.streams[0].config, selected: false },
              },
            ],
          },
        }, // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });
    const { result } = renderHook(() =>
      useCatalogTreeTableRowProps({
        ...mockStream,
        config: { selected: true, syncMode: "incremental", destinationSyncMode: "append_dedup" }, // selected true, new sync, mode and destination sync mode
      })
    );

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.added));
    // we care that iconName="plus" and className="icon plus" but it is tricky to test a react node returned from a hook like this
    expect(result.current.statusIcon).toMatchSnapshot();
    expect(result.current.pillButtonVariant).toEqual("green");
  });
  it("should return change background color with relevant icon if selected for bulk edit", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [true, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      return {
        initialValues: {
          syncCatalog: {
            streams: [
              {
                ...mockInitialValues.syncCatalog?.streams[0],
                config: { ...mockInitialValues.syncCatalog?.streams[0].config, selected: false },
              },
            ],
          },
        },
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
      } as any;
    });

    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });

    const { result } = renderHook(() =>
      useCatalogTreeTableRowProps({
        ...mockStream,
        config: { ...mockStream.config!, selected: true }, // selected true, new sync, mode and destination sync mode
      })
    );
    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.changed));
    // we care that iconName="plus" and className="icon plus" but it is tricky to test a react node returned from a hook like this
    expect(result.current.statusIcon).toMatchSnapshot();
    expect(result.current.pillButtonVariant).toEqual("blue");
  });
  it("should return change background color and override icon color with relevant icon if selected for bulk edit", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [true, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      // eslint-disable-next-line
      return { initialValues: mockInitialValues } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: undefined }] as any; // no error
    });

    const { result } = renderHook(() => useCatalogTreeTableRowProps(mockStream));

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.changed));
    // we care that iconName="plus" and className="icon plus" but it is tricky to test a react node returned from a hook like this
    expect(result.current.statusIcon).toMatchSnapshot();
    expect(result.current.pillButtonVariant).toEqual("blue");
  });
  it("should return error styles for a row that has an error", () => {
    // eslint-disable-next-line
    jest.spyOn(bulkEditService, "useBulkEditSelect").mockImplementation(() => [false, () => null] as any); // not selected for bulk edit
    jest.spyOn(connectionFormService, "useConnectionFormService").mockImplementation(() => {
      // eslint-disable-next-line
      return { initialValues: mockInitialValues } as any;
    });
    jest.spyOn(formik, "useField").mockImplementationOnce(() => {
      // eslint-disable-next-line
      return [{}, { error: true }] as any; // no error
    });

    const { result } = renderHook(() => useCatalogTreeTableRowProps(mockStream));

    expect(result.current.streamHeaderContentStyle).toEqual(classNames(styles.streamHeaderContent, styles.error));
  });
});
