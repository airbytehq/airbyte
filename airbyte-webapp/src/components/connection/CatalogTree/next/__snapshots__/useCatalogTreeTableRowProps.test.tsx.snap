// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CatalogTreeTableRow /> should return added styles for a row that is added 1`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className="icon plus"
  fade={false}
  fixedWidth={false}
  flip={null}
  icon={
    Object {
      "icon": Array [
        448,
        512,
        Array [
          10133,
          61543,
          "add",
        ],
        "2b",
        "M432 256c0 17.69-14.33 32.01-32 32.01H256v144c0 17.69-14.33 31.99-32 31.99s-32-14.3-32-31.99v-144H48c-17.67 0-32-14.32-32-32.01s14.33-31.99 32-31.99H192v-144c0-17.69 14.33-32.01 32-32.01s32 14.32 32 32.01v144h144C417.7 224 432 238.3 432 256z",
      ],
      "iconName": "plus",
      "prefix": "fas",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="2x"
  spin={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`<CatalogTreeTableRow /> should return added styles for a row that is both added and updated 1`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className="icon plus"
  fade={false}
  fixedWidth={false}
  flip={null}
  icon={
    Object {
      "icon": Array [
        448,
        512,
        Array [
          10133,
          61543,
          "add",
        ],
        "2b",
        "M432 256c0 17.69-14.33 32.01-32 32.01H256v144c0 17.69-14.33 31.99-32 31.99s-32-14.3-32-31.99v-144H48c-17.67 0-32-14.32-32-32.01s14.33-31.99 32-31.99H192v-144c0-17.69 14.33-32.01 32-32.01s32 14.32 32 32.01v144h144C417.7 224 432 238.3 432 256z",
      ],
      "iconName": "plus",
      "prefix": "fas",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="2x"
  spin={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`<CatalogTreeTableRow /> should return change background color and override icon color with relevant icon if selected for bulk edit 1`] = `null`;

exports[`<CatalogTreeTableRow /> should return change background color with relevant icon if selected for bulk edit 1`] = `null`;

exports[`<CatalogTreeTableRow /> should return removed styles for a row that is removed 1`] = `
<FontAwesomeIcon
  beat={false}
  beatFade={false}
  border={false}
  bounce={false}
  className="icon minus"
  fade={false}
  fixedWidth={false}
  flip={null}
  icon={
    Object {
      "icon": Array [
        448,
        512,
        Array [
          8722,
          10134,
          8211,
          "subtract",
        ],
        "f068",
        "M400 288h-352c-17.69 0-32-14.32-32-32.01s14.31-31.99 32-31.99h352c17.69 0 32 14.3 32 31.99S417.7 288 400 288z",
      ],
      "iconName": "minus",
      "prefix": "fas",
    }
  }
  inverse={false}
  listItem={false}
  mask={null}
  maskId={null}
  pull={null}
  pulse={false}
  rotation={null}
  shake={false}
  size="2x"
  spin={false}
  swapOpacity={false}
  symbol={false}
  title=""
  titleId={null}
  transform={null}
/>
`;

exports[`<CatalogTreeTableRow /> should return updated styles for a row that is updated 1`] = `
<div
  className="icon changed"
>
  <ModificationIcon
    color="modificationIconColor"
  />
</div>
`;
